sql: break up single-span scans into chunks

RaduBerinde opened the pull request:
In this change we break up table scans into chunks (#2966), but only when we
have a single span. We need to improve the KV API to provide a per-batch limit
when we have multiple spans (tracked by #4696).

This change will also automatically improve the LIMIT cases like the example
in #4299.

On a single node, 1,000 keys was enough to not cause performance degradation
(see below). On actual clusters we likely need bigger chunks to hide the higher
latencies; setting at 10,000 for now.

```
name \ time/op         bench-100    bench-500    bench-1000   bench-5000 bench-10000  bench-inf
Scan10000_Cockroach-4  82.5ms ±15%  42.9ms ± 4%  37.0ms ±12%  41.4ms ±46% 35.8ms ±17%  35.3ms ± 7%
```

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4759)

<!-- Reviewable:end -->

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/kvfetcher.go, line 138 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4759#-KBntKKm97xci2GVChOI:-KBntKKm97xci2GVChOJ:2078304097) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a13c5a2e778d51adf0ea48b44142e048ed48768/sql/kvfetcher.go#L138)):*
When we do a reverse scan we get the keys back in reverse order. So don't you want to be looking at `f.kvs[0]`? I've been looking at this code for a few minutes and I'm still not sure of this comment. Do we have test coverage that this is doing the right thing? I see you try different batch sizes in tests, but not where you are testing reverse scans with different batch sizes.

---

*[sql/kvfetcher.go, line 146 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4759#-KBnsbzZP_mVCkd7zjIS:-KBnsbzZP_mVCkd7zjIT:-64878151) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a13c5a2e778d51adf0ea48b44142e048ed48768/sql/kvfetcher.go#L146)):*
Did you intend to keep these commented out printfs?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4759)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/kvfetcher.go, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4759#-KBnz_Lu_mLQX3gR6dB6:-KBnz_Lu_mLQX3gR6dB7:1090671391) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a13c5a2e778d51adf0ea48b44142e048ed48768/sql/kvfetcher.go#L96)):*
instead of exporting this, export a function that takes a new value and returns a new function to restore the old, so that the caller can write `defer SetKVBatchSize(1000)()`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4759)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[sql/kvfetcher.go, line 138 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4759#-KBntKKm97xci2GVChOI:-KBo3V2vGaP97xqJ0nui:1057770042) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a13c5a2e778d51adf0ea48b44142e048ed48768/sql/kvfetcher.go#L138)):*
Ah, yes - I should make sure the test involves both types of scan, will fix.

I believe I did run all existing tests with a small batch size and they passed.

I actually wrote `f.kvs[0]` first haha. But if I request keys from `A` to`Z` in reverse order I will get `Y`, `X`, `W` .. `Q`. To resume I have to ask for keys from `A` to `Q`. It's always the last key received the one that matters (which makes sense - `f.kvs[0]` will be the same regardless how many keys I request so it can't be the one that matters).

---

*[sql/kvfetcher.go, line 146 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4759#-KBnsbzZP_mVCkd7zjIS:-KBo4FTF-JvO4X6rRmAE:-675706072) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a13c5a2e778d51adf0ea48b44142e048ed48768/sql/kvfetcher.go#L146)):*
No, forgot them in.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4759)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[sql/kvfetcher.go, line 138 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4759#-KBntKKm97xci2GVChOI:-KBoAAn9FPNRFWBGL0JA:964555165) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a13c5a2e778d51adf0ea48b44142e048ed48768/sql/kvfetcher.go#L138)):*
Updated with tests.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4759)*
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM

Let's file an issue for post-Beta to examine the performance impact of varying `kvBatchSize` settings.

---

Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/kvfetcher.go, line 138 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4759#-KBntKKm97xci2GVChOI:-KBp62Co1V2fL2vSifvQ:1241473926) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a13c5a2e778d51adf0ea48b44142e048ed48768/sql/kvfetcher.go#L138)):*
Ha, glad to hear I'm not the only one who was confused.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4759#-:-KBp6KjRx4PsM7kV1ma-:-330689349)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: break up single-span scans into chunks

In this change we break up table scans into chunks (#2966), but only when we
have a single span. We need to improve the KV API to provide a per-batch limit
when we have multiple spans (tracked by #4696).

This change will also automatically improve the LIMIT cases like the example
in #4299.

On a single node, 1,000 keys was enough to not cause performance degradation
(see below). On actual clusters we likely need bigger chunks to hide the higher
latencies; setting at 10,000 for now.
```
name \ time/op         bench-100    bench-500    bench-1000   bench-5000 bench-10000  bench-inf
Scan10000_Cockroach-4  82.5ms ±15%  42.9ms ± 4%  37.0ms ±12%  41.4ms ±46% 35.8ms ±17%  35.3ms ± 7%
```
